From 7ffece125ab4d4433fd6807292205e007a5dfd13 Mon Sep 17 00:00:00 2001
From: Kyle Stein <Kyle.Stein@arm.com>
Date: Thu, 19 Oct 2017 11:04:48 -0500
Subject: [PATCH] Increase granularity on erase process progress

This patch allows the progress bar on the LCD to be updated more
often while firmware is being erased from flash.  In certain
hardware configs, a complete erase takes several seconds and gives
the appearance that the process has stalled.  With this patch the
firmware is erased in smaller chunks with the progress bar being
updated accordingly.

Signed-off-by: Kyle Stein <Kyle.Stein@arm.com>
---
 source/active_application.cpp | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/source/active_application.cpp b/source/active_application.cpp
index 386ba74..3eee850 100755
--- a/source/active_application.cpp
+++ b/source/active_application.cpp
@@ -213,11 +213,24 @@ bool eraseActiveFirmware(uint32_t firmwareSize)
              (uint32_t) MBED_CONF_APP_FIRMWARE_METADATA_HEADER_ADDRESS + sizeRoundedUp);
 
     /* erase flash to make place for new application */
-    fota::printProgress(33, 100, "Erasing");
-    int result = flash.erase(MBED_CONF_APP_FIRMWARE_METADATA_HEADER_ADDRESS,
-                             sizeRoundedUp);
 
-    fota::printProgress(100, 100, "Erasing");
+    /* break the erase into chunks which are multiple of sector size each roughly 5%
+       of the total area to be erased */
+    uint32_t sectorsPerChunk = std::max(sizeRoundedUp / lastSectorSize / 20, (uint32_t)1);
+    uint32_t erased = 0;
+    int result = 0;
+    while ((erased < sizeRoundedUp) && (result == 0)) {
+        uint32_t chunkSize = std::min(sectorsPerChunk * lastSectorSize, sizeRoundedUp - erased);
+        result = flash.erase(MBED_CONF_APP_FIRMWARE_METADATA_HEADER_ADDRESS + erased,
+                             chunkSize);
+        if (result != 0) {
+            tr_warn("Unable to erase ActiveFirmware chunk from 0x%08" PRIX32 " to 0x%08" PRIX32,
+                    (uint32_t) MBED_CONF_APP_FIRMWARE_METADATA_HEADER_ADDRESS + erased,
+                    (uint32_t) MBED_CONF_APP_FIRMWARE_METADATA_HEADER_ADDRESS + erased + chunkSize);
+        }
+        erased += chunkSize;
+        fota::printProgress(erased, sizeRoundedUp, "Erasing");
+    }
 
     return (result == 0);
 }
-- 
2.13.5 (Apple Git-94)

